network_simulator/plot_tool.py

```diff
@@ -0,0 +1,390 @@
+            # Output average buffer occupancy to stdout.
+            PlotTool().output_sum_avg_tuple_list(
+                "Link " + link_name,
+                "Buffer Occupancy",
+                True, "(pkts)", link_stats.buffer_occupancy)
+
+            # Packet loss times.
+            plt.subplot(434)
+            loss_count_lst = PlotTool().gen_count_interval_list(
+                link_stats.packet_loss_times)
+            PlotTool().graph_tuple_list(loss_count_lst)
+            loss_legend.append(link_name)
+            # Output average packet loss per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
+                "Link " + link_name,
+                "Packet Loss", "(pkts / sec)",
+                link_stats.packet_loss_times)
+
+            # Packet transmission w.r.t time.
+            plt.subplot(437)
+            flow_rate_list = PlotTool().gen_rate_tuple_list(
+                link_stats.packet_transmit_times)
+            PlotTool().graph_tuple_list(flow_rate_list)
+            trans_legend.append(link_name)
+            # Output average packet transmitted per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
+                "Link " + link_name,
+                "Packet Transmitted", "(pkts / sec)",
+                link_stats.packet_transmit_times)
+
+        # Finalize buffer occupancy graph.
+        plt.subplot(431)
+        plt.legend(occpy_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Buffer Occupancy (pkts)")
+
+        # Finalize loss times graph.
+        plt.subplot(434)
+        plt.legend(loss_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Packet Losses")
+
+        # Finalize packet transmission graph.
+        plt.subplot(437)
+        plt.legend(trans_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Flow Rate (Mbps)")
+
+    def graph_flows(self):
+        """
+        Graphs the flow stats of each available flow.
+
+        This includes flow send and receive rates and packet RTTs.
+        """
+        # Set up legend storage for flow names.
+        # Packet sent rate legend.
+        sent_legend = []
+        # Packet received rate legend.
+        receive_legend = []
+        # RTT legend.
+        rtt_legend = []
+        # Window size legend.
+        window_legend = []
+
+        # Iterate through each flow to display each stats.
+        for flow_id, flow_stats in self.stats.flow_stats.iteritems():
+            # Name of flow would be "flow_id".
+            # Packet sent times.
+            plt.subplot(432)
+            sent_rate_lst = PlotTool().gen_rate_tuple_list(
+                flow_stats.packet_sent_times)
+            PlotTool().graph_tuple_list(sent_rate_lst)
+            sent_legend.append(flow_id)
+            # Output average packet sent per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
```

Comment:
Again, we want to compute the average send rate in Mbps, so you need to sum packet sizes in bits, divide by the largest timestamp, and then divide by MEGABIT.